hack/generated/controllers/generic_controller.go

```diff
@@ -411,6 +439,45 @@ func (gr *GenericReconciler) MonitorDeployment(ctx context.Context, action Recon
 	return result, err
 }
 
+func (gr *GenericReconciler) ManageOwnership(ctx context.Context, action ReconcileAction, data *ReconcileMetadata) (ctrl.Result, error) {
+	data.log.V(1).Info("applying ownership", "action", action)
+	isOwnerReady, err := gr.isOwnerReady(ctx, data)
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !isOwnerReady {
+		// TODO: We need to figure out how we're handing these sorts of errors.
+		// TODO: See https://github.com/Azure/k8s-infra/issues/274.
+		// TODO: For now just set an error so we at least see something
+		err := gr.Patch(ctx, data, func(ctx context.Context, mutData *ReconcileMetadata) error {
+			mutData.SetResourceError(fmt.Sprintf("owner %s is not ready", data.metaObj.Owner().Name))
+			return nil
+		})
+
+		err = client.IgnoreNotFound(err)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "patching after delete")
+		}
+
+		return ctrl.Result{
+			// TODO: We should consider a scaling backoff here, see: https://github.com/Azure/k8s-infra/issues/263
+			RequeueAfter: 5 * time.Second,
+		}, nil
+	}
+
+
+	err = gr.applyOwnership(ctx, data)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// TODO: Fast requeue as we're moving to the next stage... Do we prefer this or doing it "all at once"?
```

Comment:
We should answer this question (do we like this as a separate "state" in the state machine-thingy we have going on) or would we rather just always do this at the beginning of the `BeginDeployment` phase?

Reply:
I kind of like having a separate state for this - is the state stored on the resource? Oh, I guess it's logged by the recorder. This way if there was some bug that was preventing the owner from being ready it would be clearer why the object was stuck, rather than just being in BeginDeployment - that would make it look like maybe arm was the holdup.

Reply:
Right now there isn't a real state - it's examining a combination of properties on the resource to determine what state it's in and then acting based on that. The advantage being that we can't have drift between reported state and the properties that the state is about.

Specifically in the case of ownership, if there's no Kubernetes owner assigned to the resource and there's supposed to be (can be determined by checking `Owner()`) then we do the `ManageOwnership` step, whose output is a resource with the Kubernetes `OwnerReference` set. If you then went and manually cleared `OwnerReference` (but changed nothing else), we'd do that step again, whereas if we had a tracked state we wouldn't (because the tracked state wasn't updated back to `NeedsOwnership`) if that makes sense?

I think the notion of states is useful for us internally but may not make sense to expose to the customer (maybe events would be ok?). They're a cognitive tool to make it easier to make reconciliation eventually consistent.